synthesis/synthesis.go

```diff
@@ -260,76 +181,61 @@ func FixCds(sqlitePath string, sequence string, codontable codon.Table, problema
 				if codon.Triplet != toCodon.Triplet {
 					toCodonBias := strings.Count(toCodon.Triplet, "G") + strings.Count(toCodon.Triplet, "C")
 					switch {
-					case codonBias == toCodonBias:
-						tx.MustExec(`INSERT INTO codonbias(fromcodon, tocodon, gcbias) VALUES (?, ?, ?)`, codon.Triplet, toCodon.Triplet, "NA")
 					case codonBias > toCodonBias:
-						tx.MustExec(`INSERT INTO codonbias(fromcodon, tocodon, gcbias) VALUES (?, ?, ?)`, codon.Triplet, toCodon.Triplet, "AT")
+						atBiasMap[codon.Triplet] = append(atBiasMap[codon.Triplet], toCodon.Triplet)
 					case codonBias < toCodonBias:
-						tx.MustExec(`INSERT INTO codonbias(fromcodon, tocodon, gcbias) VALUES (?, ?, ?)`, codon.Triplet, toCodon.Triplet, "GC")
+						gcBiasMap[codon.Triplet] = append(gcBiasMap[codon.Triplet], toCodon.Triplet)
 					}
+					naBiasMap[codon.Triplet] = append(naBiasMap[codon.Triplet], toCodon.Triplet)
 				}
 			}
 		}
+		// aminoAcidTotal can't equal zero or else division later will panic
 		if aminoAcidTotal == 0 {
 			aminoAcidTotal = 1
 		}
-		weightTable[aminoAcid.Letter] = aminoAcidTotal
-	}
-
-	// Insert seq and history
-	pos := 0
-	for codonPosition := 0; codonPosition < len(sequence); codonPosition = codonPosition + codonLength {
-		codon := sequence[codonPosition : codonPosition+codonLength]
-		tx.MustExec(`INSERT INTO seq(pos) VALUES (?)`, pos)
-		tx.MustExec(`INSERT INTO history(pos, codon, step) VALUES (?, ?, 0)`, pos, codon)
-		pos++
+		aminoAcidWeightTable[aminoAcid.Letter] = aminoAcidTotal
 	}
 
+	// Build weight map
 	for _, aminoAcid := range codontable.AminoAcids {
 		for _, codon := range aminoAcid.Codons {
-			codonWeightRatio := float64(codon.Weight) / float64(weightTable[aminoAcid.Letter])
+			codonWeightRatio := float64(codon.Weight) / float64(aminoAcidWeightTable[aminoAcid.Letter])
 			normalizedCodonWeight := 100 * codonWeightRatio
-			tx.MustExec(`INSERT INTO weights(codon, weight) VALUES (?,?)`, codon.Triplet, normalizedCodonWeight)
+			weightMap[codon.Triplet] = normalizedCodonWeight
 		}
 	}
 
-	err := tx.Commit()
-	if err != nil {
-		return sequence, []Change{}, err
+	// Build historical map
+	position := 0
+	for codonPosition := 0; codonPosition < len(sequence); codonPosition = codonPosition + codonLength {
+		codon := sequence[codonPosition : codonPosition+codonLength]
+		historicalMap[position] = append(historicalMap[position], codon)
+		position++
 	}
 
 	// For a maximum of 100 iterations, see if we can do better. Usually sequences will be solved within 1-3 rounds,
 	// so 100 just effectively acts as the max cap for iterations. Once you get to 100, you pretty much know that
 	// we cannot fix the sequence.
-	for i := 1; i < fixIterations; i++ {
+	getSequence := func(history map[int][]string) string {
+		var sequence string
+		for codonPosition := 0; codonPosition < len(history); codonPosition++ {
+			sequence = sequence + history[codonPosition][len(history[codonPosition])-1]
```

Comment:
Lot of messy indexing and inline arithmetic here. Find a more verbose way of expression that is easier to read.

Reply:
I'll one up ya - how about a less verbose expression that is easier to read? Ie, use `i` instead of `codonPosition`.

This is essentially the golang equivalent of python's `my_list[-1]`, and tbh I think this "inline arithmetic" should be fairly clear.